Allow Matches and ErrorMatches matches to use an already-compiled regexp.Regexp

pajlada opened the pull request:
This allows users to have full control over the regular expression used for matching.

My use-case was matching a multiline string pattern, and the way MatchString was used in the `match`-function made that infeasible.

With this change, I was able to use the Matches matcher like this:

```go
c.Assert("asd\nasd", qt.Matches, regexp.MustCompile(`asd.*`))
```

Happy to update naming of variables on request.
I left naming of parameters/arguments on purpose to keep the diff minimal, but happy to change that too.

As an alternative, if this change is not acceptable, I know I could implement my own Checker handles regular expressions like this, so no worries if this change does not fit into the project!

frankban reviewed (commented):
I think this is an interesting feature.
If you want, please go ahead!

pajlada commented:
@frankban I've gone ahead and pushed missing tests for this functionality to maintain the 100% test coverage of `checker.go`, let me know if you prefer this to be rebased or something before being approved/merged in :slightly_smiling_face:

frankban reviewed (commented):
The code looks good!
Could you please add:
- similar tests for ErrorMatches and PanicMatches;
- updated docs for the 3 involved checkers.
Thanks!

pajlada commented:
I added the following tests:
qt.ErrorMatches single-line match `expectedNegateFailure`
qt.ErrorMatches multi-line match `expectedNegateFailure`
qt.ErrorMatches multi-line match `expectedCheckFailure`
qt.PanicMatches single-line match `expectedNegateFailure`
qt.PanicMatches multi-line match `expectedNegateFailure`
qt.PanicMatches multi-line match `expectedCheckFailure`

I added the following documentation:
Added example for Matches checker in checker.go and in README.md
Added example for ErrorMatches checker in checker.go and in README.md
Added example for PanicMatches checker in checker.go and in README.md

frankban commented:
Released as v1.14.5
